docs/trace/extending-the-sdk/README.md

```diff
@@ -334,6 +339,274 @@ Custom resource detectors can be implemented:
+
+### IServiceCollection extension methods
+
+When registering instrumentation or listening to telemetry in a library
+providing other features it is recommended to use the `IServiceCollection` as
+the target type for registration extension methods.
+
+The following example shows how a library might enable tracing and metric
+support using an `IServiceCollection` extension by calling
+`ConfigureOpenTelemetryTracing`.
+
+```csharp
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+using MyLibrary;
+using OpenTelemetry.Metrics;
+using OpenTelemetry.Trace;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class MyLibraryServiceCollectionRegistrationExtensions
+    {
+        public static IServiceCollection AddMyLibrary(
+            this IServiceCollection services,
+            string? name = null,
+            Action<MyLibraryOptions>? configure = null)
+        {
+            ArgumentNullException.ThrowIfNull(services);
+
+            // Register library services.
+            services.TryAddSingleton<IMyLibraryService, MyLibraryService>();
+
+            // Support named options.
+            name ??= Options.Options.DefaultName;
+
+            if (configure != null)
+            {
+                // Support configuration through Options API.
+                services.Configure(name, configure);
+            }
+
+            // Configure OpenTelemetry tracing.
+            services.ConfigureOpenTelemetryTracing(builder => builder.ConfigureBuilder((sp, builder) =>
+            {
+                var options = sp.GetRequiredService<IOptionsMonitor<MyLibraryOptions>>().Get(name);
+                if (options.EnableTracing)
+                {
+                    builder.AddSource("MyLibrary");
+                }
+            }));
+
+            // Configure OpenTelemetry metrics.
+            services.ConfigureOpenTelemetryMetrics(builder => builder.ConfigureBuilder((sp, builder) =>
+            {
+                var options = sp.GetRequiredService<IOptionsMonitor<MyLibraryOptions>>().Get(name);
+                if (options.EnableMetrics)
+                {
+                    builder.AddMeter("MyLibrary");
+                }
+            }));
+
+            return services;
+        }
+    }
+}
+
+namespace MyLibrary
+{
+    // Options class can be bound to IConfiguration or configured by code.
+    public class MyLibraryOptions
+    {
+        public bool EnableTracing { get; set; }
+
+        public bool EnableMetrics { get; set; }
+    }
+
+    internal sealed class MyLibraryService : IMyLibraryService
+    {
+        // Implementation not shown.
+    }
+
+    public interface IMyLibraryService
+    {
+        // Implementation not shown.
+    }
+}
+```
+
+The benefit to using the `IServiceCollection` style is users only need to call a
+single `AddMyLibrary` extension to configure the library itself and optionally
+turn on OpenTelemetry integration.
```

Comment:
This could be said about the `TracerProviderBuilder` style as well. I think the thing to highlight here about the `IServiceCollection` style is just what your example demonstrates - as an instrumentation author, I'm likely only taking an API dependency not an SDK dependency, so therefore the guidance you provide me is to offer an extension method on `IServiceCollection` making it easy to configure the instrumentation.

Reply:
🤷 here's a stab at this

> One of the main benefits of the `IServiceCollection` style is that it provides a way for library authors, taking only a dependency on the OpenTelemetry API, to offer an extension for configuring telemetry emitted from the library. This is commonly the case for instrumentation library authors.

Reply:
I tweaked this a bit to clarify my point and merged so I could keep moving on the configuration section. But feel free to modify further, I'm not at all sure it is perfect yet 😄

> taking only a dependency on the OpenTelemetry API

At the moment, to use this stuff you MUST have an SDK reference. If you only have an API reference, you can do `ConfigureBuilder` types of things by using `IDeferredTracerProviderBuilder`. But there is nothing in API to give you access to the `IServiceCollection` because API doesn't have a Microsoft.Extensions.DependencyInjection.Abstractions reference (it can't see `ISeviceCollection`). That's what this PR was doing: https://github.com/open-telemetry/opentelemetry-dotnet/pull/3663. If we add that DependencyInjection.Abstractions reference to API, we can move some of this fundamental stuff into API and enable this scenario with that lower reference.

Reply:
> At the moment, to use this stuff you MUST have an SDK reference.

Ha, yup. We've been talking about all this stuff for too long 😆. My brain wires are crossed between reality and things that have simply been proposed.